### PR TITLE
Added hyperparameter tuning example.

### PR DESCRIPTION
**Pull request by tomasatdatabricks**
Example of how you can use hyper param tuning with mlflow using external hyper param tuning libraries.  

examples/hyperparam/MLProject has 4 targets:
  - main 
     trains simple deep learning model on the wine-quality dataset from our tutorial.
     Has 2 tunable parameters - learning_rate and momentum
 - grid
    simple grid search over learning rate and momentum parameters
- gpyopt 
   uses GPyOpt library to optimize hyper parameters for main
- HyperOpt
    uses Hyperopt library to optimize hyper parameters for main

Both targets take optional expeirment id for training runs. If provided, training runs will be logged under this experiment id (to avoid confusion in the ui).

Example usage:

mlflow run  -e hyperopt --experiment-id 5 -P max_runs=16 -P max_epochs=32  -P training_experiment_id=6 example/hyperparam

### CODE REVIEW

File: example/hyperparam/MLproject [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                             --max-runs {max_runs} \
+                                             --batch-size {batch_size} \
+                                             --max-p {max_p} \
+                                             --epochs {max_epochs} \
+                                             --metric {metric}
+                                             --gpy-model {gpy_model} \
+                                             --gpy-acquisition {gpy_acquisition} \
+                                             --initial-design {initial_design} \
+                                             --seed {seed} \
+                                             --training-experiment-id {training_experiment_id}"
```
Comment (mateiz) [inline example/hyperparam/MLproject]: Nit: how come some of these multiline strings have backslashes and others don't?

File: example/hyperparam/README.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
+Hyper Parameter Tuning Example
```
Comment (mateiz) [inline example/hyperparam/README.rst]: Nit: I think we should say "hyperparameter" (one word) throughout. Right now we have a mix of things: hyper parameter, hyperparameter, hyper param, etc. Just stick with one.

File: example/hyperparam/README.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+------------------------------
+
+Example of how you can use hyper param tuning with mlflow using external hyper param tuning
```
Comment (mateiz) [inline example/hyperparam/README.rst]: Nit: capitalize MLflow, Keras, DL, etc consistently throughout.

File: example/hyperparam/README.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+libraries. Both examples are implemented as an MLflow run entry point which evaluates the model by
+calling another MLflow run entry point. This way both the parent hyper param tuning run and the
+spawned training runs get logged. Both targets take optional experiment id for training runs. If
+provided, training runs will be logged under this experiment id. This is a short term solution to
+organizing the runs so that it is easy to view individual training runs and the hyper param runs
+separately. In the future this will be achieved by MLflow tags.
+
+examples/hyperparam/MLproject has 3 targets:
+  * dl_train
```
Comment (mateiz) [inline example/hyperparam/README.rst]: Maybe we should just call this "train" to make it consistent with the others

File: example/hyperparam/README.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    trains simple deep learning model on the wine-quality dataset from our tutorial.
+    It has 3 tunable hyper parameters - learning_rate, beta1, beta2.
+    Contains examples of how keras callbacks can be used for mlflow integration.
+  * GPyOpt
+    uses `GPyOpt <https://github.com/SheffieldML/GPyOpt>`_ to optimize hyper parameters of dl_train
+    GPyOpt can run multiple mlflow runs in parallel if run with batch-size > 1 and max_p > 1.
+  * HyperOpt
```
Comment (mateiz) [inline example/hyperparam/README.rst]: Maybe these entry point names should be lowercase to match our other examples

File: example/hyperparam/README.rst [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+.. code:: bash
+
+    mlflow experiments create hyper_param_runs
+
+This will create experiment for hyper param runs and return its experiment it.
+
+.. code:: bash
+
+    mlflow run  -e dl_train --experiment-id <individual_runs_experiment_id> example/hyperparam
```
Comment (mateiz) [inline example/hyperparam/README.rst]: Maybe we only want one space between `run` and `-e`? Or was something supposed to go there?

File: example/hyperparam/conda.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+channels:
+  - defaults
+dependencies:
+  - python=3.6
+  - numpy=1.14.3
+  - pandas=0.22.0
+  - scikit-learn=0.19.1
+  - matplotlib=2.2.2
+  - pip:
+    - mlflow==0.5.1
```
Comment (mateiz) [inline example/hyperparam/conda.yaml]: Might be good to do `>=` here or not specify a version since this code will remain in the repo.

File: example/hyperparam/gPyOpt/search_keras.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
+import math
```
Comment (mateiz) [inline example/hyperparam/gPyOpt/search_keras.py]: Why is this called search_keras.py instead of just search.py or gpyopt.py? It's not specific to Keras really. I'd just call it gpyopt.py and rename the HyperOpt one in a similar way to make them easier to find.

File: example/hyperparam/gPyOpt/search_keras.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    bounds = [
+        {'name': 'lr', 'type': 'continuous', 'domain': (1e-3, 1e-1)},
+        {'name': 'lr_decay', 'type': 'continuous', 'domain': (.8, 1)},
+        {'name': 'rho', 'type': 'continuous', 'domain': (.8, 1)},
+    ]
+    # create random file to store run ids of the training tasks
+    tmp = tempfile.mkdtemp()
+    results_path = os.path.join(tmp, "results.txt")
+
+    def eval(parms):
```
Comment (mateiz) [inline example/hyperparam/gPyOpt/search_keras.py]: Maybe "parms" should be "params"?

File: example/hyperparam/gPyOpt/search_keras.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        Train keras model with given parameters by invoking mlflow run.
+
+        Notice we store runUuid and resulting metric in a file. We will later use these to pick the
+        best run and to log the runUuids of the child runs as an artifact.
```
Comment (mateiz) [inline example/hyperparam/gPyOpt/search_keras.py]: Maybe we can write that this is a temporary workaround until we can add tags to our MLflow runs to search for them directly.

File: example/hyperparam/hyperopt/search_keras.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
+import click
```
Comment (mateiz) [inline example/hyperparam/hyperopt/search_keras.py]: Add a small doc comment at the top of each file to say what it does. It makes it easier to understand than scrolling down into the file. See some of the other examples for reference.

File: example/hyperparam/hyperopt/search_keras.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+@click.argument("training_data")
+def train(training_data, max_runs, epochs, metric, algo, seed, training_experiment_id):
+    """
+    Run hyper param optimization.
+    """
+    # create random file to store run ids of the training tasks
+    tmp = tempfile.mkdtemp()
+    results_path = os.path.join(tmp, "results")
+
+    def eval(parms):
```
Comment (mateiz) [inline example/hyperparam/hyperopt/search_keras.py]: Same thing, I'd say "params" instead of "parms" to be more standard.

File: example/hyperparam/train_keras.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        self._score_f = None
+
+    def on_epoch_end(self, epoch, logs=None):
+        # log standard keras metrics
+        if logs:
+            for k, v in logs.items():
+                mlflow.log_metric(k, v)
+
+
+class MLflowCheckpoint(Callback):
```
Comment (mateiz) [inline example/hyperparam/train_keras.py]: Add a doc comment for this; it's cool but we should tell people why we added this

File: example/hyperparam/train_keras.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+def get_standardize_f(train):
+    mu = np.mean(train, axis=0)
+    std = np.std(train, axis=0)
+    return lambda x: (x - mu) / std
+
+
+from keras.callbacks import Callback
+
+
+class MLflowLogger(Callback):
```
Comment (mateiz) [inline example/hyperparam/train_keras.py]: Add a doc comment for this; it's cool but we should tell people why we added this

File: example/hyperparam/train_keras.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    test_y = test[["quality"]].as_matrix().astype("float32")
+
+    with mlflow.start_run():
+        rmse, mae, r2 = eval_metrics(test_y, np.ones(len(test_y)) * np.mean(test_y))
+        mlflow.log_metric("rmse_null", rmse)
+        mlflow.log_metric("mae_null", mae)
+        mlflow.log_metric("r2_null", r2)
+        print("  RMSE (null): %s" % rmse)
+        print("  MAE (null) : %s" % mae)
+        print("  R2 (null): %s" % r2)
```
Comment (mateiz) [inline example/hyperparam/train_keras.py]: Why do we want to log these separately?

**Review by mateiz — COMMENTED** [review]
Left a few comments, many of which are about adding more docs. I haven't tried running it yet but I'll leave more once I do that.

File: example/hyperparam/train_keras.py (hunk 15 above)
Comment (tomasatdatabricks) [reply example/hyperparam/train_keras.py]: Good question. 2 reasons:

So that the caller can use the null model metric as max reasonable value. This is needed in the GpyOpt because some configurations fail or return NaNs or Inf. In that case I need to be able to fill in reasonable value (math.inf does not work unfortunately) which is >= all other values.

I think it is always useful to compare the metric against the value obtained from the null model.

I could always log the metric of the null model as the first value of the metric, but then the user might need to get the history and take the first value which seems more error prone.

File: example/hyperparam/train_keras.py (hunk 14 above)
Comment (tomasatdatabricks) [reply example/hyperparam/train_keras.py]: Makes sense, will do!

File: example/hyperparam/MLproject (hunk 1 above)
Comment (tomasatdatabricks) [reply example/hyperparam/MLproject]: Good question, I guess they are not really needed. I did not realize. I will remove them.

File: example/hyperparam/README.rst (hunk 2 above)
Comment (tomasatdatabricks) [reply example/hyperparam/README.rst]: Makes sense.

File: example/hyperparam/search_gpyopt.py [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+                                                                    math.inf,
+                                                                    True)(params=[[0, 0]])
+        myProblem = GPyOpt.methods.BayesianOptimization(new_eval(epochs,
+                                                                 experiment_id,
+                                                                 train_null_loss,
+                                                                 valid_null_loss,
+                                                                 test_null_loss),
+                                                        bounds,
+                                                        batch_size=batch_size,
+                                                        num_cores=max_p,
```
[the remark this answers is not in the document]
Comment (mateiz) [reply example/hyperparam/search_gpyopt.py]: BTW this was with batch_size=4 and max_p=4. There were two parallel runs at the start but then everything is serial.

[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply example/hyperparam/search_gpyopt.py]: That is because the default evaluator for GPyOpt is sequential. This means that the first n random samples are evaluated in parallel but after that the models are evaluated one by one. I will change this to evaluator which works in parallel.

File: examples/hyperparam/search_gpyopt.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+                    "momentum": str(momentum),
+                    "seed": str(seed)},
+                experiment_id=experiment_id,
+                block=False
+            )
+            store = mlflow.tracking._get_store()
+            if p.wait():
+                # cap the loss at the loss of the null model
+                train_loss = min(null_valid_loss,
+                                 store.get_metric(p.run_id, "train_{}".format(metric)).value)
```
Comment (aarondav) [inline examples/hyperparam/search_gpyopt.py]: @tomasatdatabricks We were trying to find a good example that uses the public APIs to get metrics from past runs, and this looks great except that `store.get_metric` is not a public API. Could we use get_run() on the Tracking Service API and parse the metrics out from there? I know it won't look as pretty, but that means we should think about adding a pretty public API rather than having an example with an internal API.

Comment (tomasatdatabricks) [reply examples/hyperparam/search_gpyopt.py]: Sure, will do, it makes sense to have public apis. We could definitely improve the API here.

File: examples/hyperparam/search_random.py [hunk 18]
```diff
@@ -67,10 +66,16 @@ def eval(parms):
                 experiment_id=experiment_id,
                 block=False)
             if p.wait():
+                training_run = tracking_service.get_run(p.run_id)
+
+                def get_metric(metric_name):
+                    return [m.value for m in training_run.data.metrics if m.key == metric_name][0]
+
+
```
Comment (aarondav) [inline examples/hyperparam/search_random.py]: nit: extra newline

**Comment by aarondav** [comment]
Thanks @tomasatdatabricks -- updated metric reading lgtm and this is now a good example of reading historical metrics using the tracking api.

File: examples/hyperparam/README.rst [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+  * main
+    train simple deep learning model on the wine-quality dataset from our tutorial.
+    It has 3 tunable hyperparameters - learning_rate, beta1, beta2.
+    Contains examples of how Keras callbacks can be used for mlflow integration.
+  * random
+    perform simple random search over the parameter space.
+  * gpyopt
+    use `GPyOpt <https://github.com/SheffieldML/GPyOpt>`_ to optimize hyperparameters of train.
+    GPyOpt can run multiple mlflow runs in parallel if run with batch-size > 1 and max_p > 1.
+  * hyperOpt
```
Comment (mateiz) [inline examples/hyperparam/README.rst]: Seems to be called `hyperopt`, not `hyoerOpt`.

File: examples/hyperparam/README.rst [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+.. code:: bash
+
+    mlflow experiments create hyper_param_runs
+
+This will create experiment for hyperparam runs and return its experiment it.
+
+.. code:: bash
+
+    mlflow run  --experiment-id <individual_runs_experiment_id> example/hyperparam
```
Comment (mateiz) [inline examples/hyperparam/README.rst]: Nit: for some reason you have two spaces after "run" here and in other commands below, instead of one.

File: examples/hyperparam/search_gpyopt.py [hunk 21]
```diff
@@ -0,0 +1,212 @@
+"""
+Example of hyperparameter search in MLflow using GPyOpt.
+
+The run method will instantiate and run gpyopt optimizer. Each parameter configuration is
+evaluated in a new mlflow run invoking main entry point with selected parameters.
```
Comment (mateiz) [inline examples/hyperparam/search_gpyopt.py]: Capitalize GPyOpt, MLflow etc consistently throughout the docs.

File: examples/hyperparam/search_hyperopt.py [hunk 22]
```diff
@@ -0,0 +1,181 @@
+"""
+Example of hyperparameter search in MLflow using Hyperopt.
+
+The run method will instantiate and run hyperopt optimizer. Each parameter configuration is
+evaluated in a new mlflow run invoking main entry point with selected parameters.
```
Comment (mateiz) [inline examples/hyperparam/search_hyperopt.py]: Capitalize Hyperopt and MLflow consistently

File: examples/hyperparam/search_hyperopt.py [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+        :valid_null_loss: Loss of a null model on the validation dataset
+        :test_null_loss: Loss of a null model on the test dataset.
+        :return_test_loss: Return both validation and test loss if set.
+
+        :return: new eval function.
+        """
+
+        def eval(params):
+            """
+            Train keras model with given parameters by invoking mlflow run.
```
Comment (mateiz) [inline examples/hyperparam/search_hyperopt.py]: Capitalize Keras consistently in docs

File: examples/hyperparam/search_random.py [hunk 24]
```diff
@@ -0,0 +1,136 @@
+"""
+Example of hyperparameter search in MLflow using simple random search.
+
+The run method will evaluate random combinations of parameters in a new mlflow run.
```
Comment (mateiz) [inline examples/hyperparam/search_random.py]: MLflow

File: examples/hyperparam/train.py [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+from keras.models import Sequential
+from keras.layers import Dense, Lambda
+from keras.optimizers import SGD
+from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
+from sklearn.model_selection import train_test_split
+
+import mlflow
+import mlflow.keras
+
+"""
```
Comment (mateiz) [inline examples/hyperparam/train.py]: Move this comment to the top of the file so it's easier to find; this is a doc comment for the whole file, not a method or something

**Review by mateiz — COMMENTED** [review]
Left some more minor notes on the docs
